File: leetcode/01_alg/13_trie_disjoint_set/disjoint_set/200.岛屿数量.py
"""
 * @lc app=leetcode.cn id=200 lang=cpp
 *
 * [200] 岛屿数量
 *
 * https://leetcode.cn/problems/number-of-islands/description/
 *
 * algorithms
 * Medium (59.68%)
 * Likes:    2336
 * Dislikes: 0
 * Total Accepted:    720.8K
 * Total Submissions: 1.2M
 * Testcase Example:  '[["1","1","1","1","0"],["1","1","0","1","0"],["1","1","0","0","0"],["0","0","0","0","0"]]'
 *
 * 给你一个由 '1'（陆地）和 '0'（水）组成的的二维网格，请你计算网格中岛屿的数量。
 * 
 * 岛屿总是被水包围，并且每座岛屿只能由水平方向和/或竖直方向上相邻的陆地连接形成。
 * 
 * 此外，你可以假设该网格的四条边均被水包围。
 * 
 * 
 * 
 * 示例 1：
 * 
 * 
 * 输入：grid = [
 * ⁠ ["1","1","1","1","0"],
 * ⁠ ["1","1","0","1","0"],
 * ⁠ ["1","1","0","0","0"],
 * ⁠ ["0","0","0","0","0"]
 * ]
 * 输出：1
 * 
 * 
 * 示例 2：
 * 
 * 
 * 输入：grid = [
 * ⁠ ["1","1","0","0","0"],
 * ⁠ ["1","1","0","0","0"],
 * ⁠ ["0","0","1","0","0"],
 * ⁠ ["0","0","0","1","1"]
 * ]
 * 输出：3
 * 
 * 
 * 
 * 
 * 提示：
 * 
 * 
 * m == grid.length
 * n == grid[i].length
 * 1 
 * grid[i][j] 的值为 '0' 或 '1'
 * 
 * 
 */
"""

# @lc code=start
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        """
        为了求出岛屿的数量，可以扫描整个二维网格。如果一个位置为 1，
        则将其与相邻四个方向上的 111 在并查集中进行合并。

        最终岛屿的数量就是并查集中连通分量的数目。
        """
        if not grid:
            return 0
        
        m = len(grid)
        n = len(grid[0])
        # 初始化并查集
        self.parent = [i for i in range(m * n)]
        self.count =0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    self.parent[i * n + j] = i * n + j
                    self.count += 1

        # 合并
        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    grid[i][j] = "0"
                    for k in range(len(dx)):
                        x, y = i + dx[k], j + dy[k]
                        if 0 <= x < m and 0 <= y < n and grid[x][y] == "1":
                            self._union(i * n + j, x * n + y)

        return self.count

    def _union(self, i, j):
        p1 = self._parent(i)
        p2 = self._parent(j)
        if p1 == p2:
            return
        self.parent[p1] = p2
        self.count -= 1

    def _parent(self, i):
        root = i
        while self.parent[root] != root:
            root = self.parent[root]
        # 路径压缩
        while self.parent[i] != i:
            x = i
            i = self.parent[i]
            self.parent[x] = root
        return root
# @lc code=end

File: leetcode/03_algo_note/02_basic_ds/2.6_dfs/200.岛屿数量.cpp
/*
 * @lc app=leetcode.cn id=200 lang=cpp
 *
 * [200] 岛屿数量
 *
 * https://leetcode.cn/problems/number-of-islands/description/
 *
 * algorithms
 * Medium (59.95%)
 * Likes:    2396
 * Dislikes: 0
 * Total Accepted:    749.9K
 * Total Submissions: 1.3M
 * Testcase Example:  '[["1","1","1","1","0"],["1","1","0","1","0"],["1","1","0","0","0"],["0","0","0","0","0"]]'
 *
 * 给你一个由 '1'（陆地）和 '0'（水）组成的的二维网格，请你计算网格中岛屿的数量。
 * 
 * 岛屿总是被水包围，并且每座岛屿只能由水平方向和/或竖直方向上相邻的陆地连接形成。
 * 
 * 此外，你可以假设该网格的四条边均被水包围。
 * 
 * 
 * 
 * 示例 1：
 * 
 * 
 * 输入：grid = [
 * ⁠ ["1","1","1","1","0"],
 * ⁠ ["1","1","0","1","0"],
 * ⁠ ["1","1","0","0","0"],
 * ⁠ ["0","0","0","0","0"]
 * ]
 * 输出：1
 * 
 * 
 * 示例 2：
 * 
 * 
 * 输入：grid = [
 * ⁠ ["1","1","0","0","0"],
 * ⁠ ["1","1","0","0","0"],
 * ⁠ ["0","0","1","0","0"],
 * ⁠ ["0","0","0","1","1"]
 * ]
 * 输出：3
 * 
 * 
 * 
 * 
 * 提示：
 * 
 * 
 * m == grid.length
 * n == grid[i].length
 * 1 
 * grid[i][j] 的值为 '0' 或 '1'
 * 
 * 
 */

// @lc code=start
class Solution {
public:
    // 填海造陆
    // 遍历地图，遇到一个1，将1进行BFS/DFS遍历，将周围的1全部变为0，操作次数记为一次
    // 操作次数就是岛屿数量
    int numIslands(vector<vector<char>>& grid) {
        int island_num = 0;
        m_grid = grid;

        for (int i = 0; i < m_grid.size(); i++) {
            for (int j = 0; j < m_grid[i].size(); j++) {
                if (m_grid[i][j] == '0') {
                    continue;
                }
                island_num += this->sink(i, j);
            }
        }

        return island_num;
    }
private:
    // 方向
    int m_dx[4] = {-1, 1, 0, 0};
    int m_dy[4] = {0, 0, -1, 1};
    // 全局地图
    std::vector<std::vector<char>> m_grid;

    // bfs遍历
    int sink(int i, int j) {
        if (m_grid[i][j] == '0') {
            return 0;
        }

        // 将i，j变为0
        m_grid[i][j] = '0';
        // 开始遍历四个方向
        for (int k = 0; k < 4; k++) {
            int x = i + m_dx[k];
            int y = j + m_dy[k];

            if (x >=0 && x < m_grid.size() && y >= 0 && y < m_grid[i].size()) {
                // 遇到0跳过
                // 遇到1开始递归
                if (m_grid[x][y] == '0') {
                    continue;
                }
                this->sink(x, y);
            }
        }
        return 1;
    }
};
// @lc code=end

